unit/mr_test.c

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright (c) 2013-2016 Intel Corporation.  All rights reserved.
+ * Copyright (c) 2014-2016 Cisco Systems, Inc.  All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AWV
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <getopt.h>
+#include <string.h>
+
+#include <rdma/fi_errno.h>
+
+#include "unit_common.h"
+#include "shared.h"
+
+static char err_buf[512];
+
+/*
+ * Tests:
+ */
+static int mr_reg()
+{
+	int i;
+	int ret;
+	int testret;
+	struct fid_mr *mr;
+
+	testret = FAIL;
+
+	for (i = 0; i < test_cnt; i++) {
+		buf_size = test_size[i].size;
+		buf = malloc(buf_size);
+		if (!buf) {
+			ret = -FI_ENOMEM;
+			goto fail;
+		}
+
+		// TODO test for various access
+		ret = fi_mr_reg(domain, buf, buf_size, FI_SEND | FI_RECV, 0,
```

Comment:
Registering for SEND/RECV is only necessary for FI_LOCAL_MR.  For example, the UDP provider doesn't support memory registration, since it doesn't need it.  I suggest setting a global variable for access based on the output from fi_getinfo.

We may need to call fi_getinfo multiple times until we either get a supported registration mode, or determine that it doesn't apply.  For example, first call asking for RMA access; if yes, set access = REMOTE_READ/WRITE.  If nothing, call with FI_LOCAL_MR mode supported; if required, set access = SEND/RECV.  Otherwise just return success.  (I'm not sure this covers all cases, but hopefully you get the idea.)